symbolic-base/src/ZkFold/Algebra/Polynomial/Univariate/Simple.hs

```diff
@@ -0,0 +1,108 @@
+-- since its operations do not require 'Prelude.Eq'.
+newtype SimplePoly a n = SimplePoly
+    { coeffs :: V.Vector a
+      -- ^ Vector of coefficients in ascending-degree order.
+      --
+      -- NOTE there is no guarantee that length of 'coeffs' is @n@,
+      -- only that its length is _not greater than_ @n@.
+      -- To get correctly sized vector, use 'toVector'.
+      --
+      -- NOTE since 'SimplePoly' does not use 'Prelude.Eq', it is possible that
+      -- there are some non-zeroed greater coefficients. Use with care.
+    }
+
+-- | Vector of exactly @n@ coefficients in ascending-degree order.
+toVector ::
+    forall a n . (AdditiveMonoid a, KnownNat n) =>
+    SimplePoly a n -> ZkFold.Vector n a
+toVector (SimplePoly cs) =
+    ZkFold.Vector $ cs V.++ V.replicate (integral @n Int.- V.length cs) zero
+
+instance {-# OVERLAPPING #-} FromConstant (SimplePoly a n) (SimplePoly a n)
+
+instance FromConstant k a => FromConstant k (SimplePoly a n) where
+    fromConstant = SimplePoly . V.singleton . fromConstant
+
+instance {-# OVERLAPPING #-}
+    (Semiring a, KnownNat n) => Scale (SimplePoly a n) (SimplePoly a n)
+
+instance Scale k a => Scale k (SimplePoly a n) where
+    scale k = SimplePoly . fmap (scale k) . coeffs
+
+instance (Semiring a, KnownNat n) => Exponent (SimplePoly a n) Natural where
+    (^) = natPow
+
+instance AdditiveSemigroup a => AdditiveSemigroup (SimplePoly a n) where
+    SimplePoly p + SimplePoly q = SimplePoly $ alignWith (T.mergeThese (+)) p q
+
+instance AdditiveMonoid a => AdditiveMonoid (SimplePoly a n) where
+    zero = SimplePoly V.empty
+
+instance AdditiveGroup a => AdditiveGroup (SimplePoly a n) where
+    negate = SimplePoly . fmap negate . coeffs
+
+instance
+    (Semiring a, KnownNat n) => MultiplicativeSemigroup (SimplePoly a n) where
+    SimplePoly p * SimplePoly q =
+        let pLen = V.length p
+            qLen = V.length q
+            pqLen = case pLen Int.+ qLen of
+                        0 -> 0
+                        l -> min (integral @n) (l Int.- 1)
+         in SimplePoly $ V.generate pqLen \k -> sum [
+            (p V.! i) * (q V.! j)
+            | i <- [0 .. min pLen k], let j = k Int.- i, j < qLen
+         ]
+
+instance (Semiring a, KnownNat n) => MultiplicativeMonoid (SimplePoly a n) where
+    one = fromConstant (one :: a)
+
+instance (Semiring a, KnownNat n) => Semiring (SimplePoly a n)
+
+instance (Ring a, KnownNat n) => Ring (SimplePoly a n)
+
+instance Ring a => UnivariateRingPolyVec a (SimplePoly a) where
+    SimplePoly p .*. SimplePoly q =
+        SimplePoly $ alignWith (uncurry (*) . T.fromThese zero zero) p q
+    (.*) = flip scale
+    (*.) = scale
+    (+.) c = SimplePoly . fmap (c +) . coeffs
+    (.+) = flip (+.)
+    toPolyVec :: forall n . KnownNat n => Vector a -> SimplePoly a n
+    toPolyVec = SimplePoly . V.take (integral @n)
+    fromPolyVec = coeffs
+    poly2vec = toPolyVec . fromPoly
+    vec2poly = toPoly . fromPolyVec
```

Comment:
I feel like for `poly2vec` and `vec2poly`, there's exactly one implementation thanks to their type: they have to work with arbitrary `Polynomial`s